core/modules/css/KvInterfaceImplMySql.cc

```
@@ -0,0 +1,503 @@
+    KvTransaction transaction(_conn);
+
+    std::string val;
+    sql::SqlErrorObject errObj;
+
+    boost::format fmQuery("SELECT kvVal FROM kvData WHERE kvKey='%1%'");
+    fmQuery % key;
+    std::string query = fmQuery.str();
+
+    sql::SqlResults results;
+    if (not _conn.runQuery(query, results, errObj)) {
+        LOGF(_logger, LOG_LVL_ERROR, "_get - query failed: %1%" % query);
+        if (throwIfKeyNotFound) {
+            throw CssError(errObj);
+        } else {
+            val = defaultValue;
+        }
+    } else {
+        errObj.reset();
+        if (not results.extractFirstValue(val, errObj)) {
+            if (throwIfKeyNotFound) {
+                throw NoSuchKey(errObj);
+            } else {
+                val = defaultValue;
+            }
+        }
+    }
+
+    transaction.commit();
+    return val;
+}
+
+
+bool KvInterfaceImplMySql::_getIdFromServer(std::string const& key, unsigned int* id) {
+    // a transaction should be in process here.
+
+    boost::format fmQuery("SELECT kvId FROM kvData WHERE kvKey='%1%'");
+    fmQuery % key;
+    std::string query = fmQuery.str();
+
+    sql::SqlResults results;
+    sql::SqlErrorObject errObj;
+    if (not _conn.runQuery(query, results, errObj)) {
+        LOGF(_logger, LOG_LVL_ERROR, "_getIdFromServer - query failed: %1%" % query);
+        return false;
+    }
+    errObj.reset();
+
+    if (not extractIntValueFromSqlResults(&results, id)) {
+        return false;
+    }
+
+    return true;
+}
+
+
+std::string KvInterfaceImplMySql::_validateKey(std::string const& key) {
+    // There is no need for a transaction here.
+
+    sql::SqlErrorObject errObj;
+    std::string retStr;
+    if (not _conn.escapeString(key, retStr, errObj)) {
+        throw CssError(errObj);
+    }
+    // verify that:
+    // - key is less than max length
+    // - first character should be a delimiter
+    // - last character should not be a delimiter
+    if (key.size() > KvInterface::MAX_KEY_LENGTH ||
```

mysql is very ugly in that respect - by default it silently truncates long strings. We should probably use non-default option to force it generate errors in this case. I'd still keep that check.